### PR TITLE
add type promotion logic for eager between tensor and tensor

### PR DESCRIPTION
**Pull request by zxcd**
### PR types

New features

### PR changes

Others

### Description

Card-64022
add type promotion logic for eager between tensor and tensor.

stage 1 will only support float in "+ - *, where" API.
Example like:
```
>>> a = paddle.randn((2,3), dtype='float32')
>>> b = paddle.randn((2,3), dtype='float64')
>>> out = a + b

# Before, dtype of out would be same to the former one, i.e. a
>>> out.dtype
<VarType.FP32: 5>

# Now, dtype will be promote to float64
>>> out
<VarType.FP64: 6>
```

### CODE REVIEW

File: test/legacy_test/test_math_op_patch_var_base.py [hunk 1]
```diff
@@ -53,6 +192,75 @@ def test_mul(self):
...
+        a_np = np.random.random(self.shape).astype(np.float16)
+        b_np = np.random.random(self.shape).astype(np.float32)
+        b_np = convert_uint16_to_float(convert_float_to_uint16(b_np))
+        with base.dygraph.guard():
+            a = base.dygraph.to_variable(a_np)
+            b = base.dygraph.to_variable(b_np).astype('bfloat16')
+            res = a * b
+            res_t = b * a
+            np.testing.assert_array_equal(res_t.numpy(), res.numpy())
+            np.testing.assert_array_equal(res.numpy(), a_np * b_np)
```
Comment (zoooo0820) [inline test/legacy_test/test_math_op_patch_var_base.py:270]: Can these test cases detect the Dtype of output be as expected?

File: test/legacy_test/test_where_op.py [hunk 2]
```diff
@@ -457,6 +457,141 @@ def test_dygraph_api_broadcast_8(self):
...
+                    paddle.to_tensor(y),
+                ).numpy(),
+            )
+            np.testing.assert_array_equal(
+                np.where(cond, y, x),
+                paddle.where(
+                    paddle.to_tensor(cond),
+                    paddle.to_tensor(y),
+                    paddle.to_tensor(x),
+                ).numpy(),
```
Comment (zoooo0820) [inline test/legacy_test/test_where_op.py:487]: same as above, should we add a case to test the expected dtype

File: test/legacy_test/test_math_op_patch_var_base.py [hunk 3]
```diff
@@ -35,6 +36,87 @@ def test_add(self):
...
+            res_t = b + a
+            np.testing.assert_equal(paddle.float32, res.dtype)
+            np.testing.assert_equal(res_t.dtype, res.dtype)
+            np.testing.assert_array_equal(res_t.numpy(), res.numpy())
+            np.testing.assert_array_equal(res.numpy(), a_np + b_np)
+
+    def test_type_promotion_add_F4_BF(self):
+        a_np = np.random.random(self.shape).astype(np.float32)
+        b_np = np.random.random(self.shape).astype(np.float32)
+        b_np = convert_uint16_to_float(convert_float_to_uint16(b_np))
```
Comment (zoooo0820) [inline test/legacy_test/test_math_op_patch_var_base.py:95]: the dtype convertion for `b_np` seems unnecessary

Comment (zxcd) [reply test/legacy_test/test_math_op_patch_var_base.py:95]: b_np convert for fake a np.bfloat16.

**Review by jeff41404 — DISMISSED** [review]
LGTM

File: paddle/fluid/eager/auto_code_generator/generator/eager_gen.py [hunk 4]
```diff
@@ -1804,7 +1835,35 @@ def GenerateForwardDefinitionAndDeclaration(self, is_inplaced):
...
+            type_promote_list_str = f"auto new_{x} = egr::PromoteCast(\"{x}\", {x}, promotion_type);\n"
+            type_promote_list_str += f"    auto new_{y} = egr::PromoteCast(\"{y}\", {y}, promotion_type);\n"
 
+            type_promotion_logic_str = TYPE_PROMOTION_LOGIC_TEMPLATE.format(
+                need_type_promote_str,
+                kernel_trans2_op_name_str,
+                type_promote_get_dst_dtype_str,
+                type_promote_list_str,
+                type_promote_call_list,
+            )
```
Comment (zyfncg) [inline paddle/fluid/eager/auto_code_generator/generator/eager_gen.py]: 这里是否可以化简下，仅将`x`, `y`,`forward_ad_function_name`等少量参数作为`TYPE_PROMOTION_LOGIC_TEMPLATE`的输入参数，这样代码可读性更好些，可以参考
image

Comment (zxcd) [reply paddle/fluid/eager/auto_code_generator/generator/eager_gen.py]: 好的 我修改一下

**Review by zyfncg — APPROVED** [review]

**Review by jeff41404 — APPROVED** [review]

**Pull request merged**